cppcache/integration/test/PdxSerializableTest.cpp

```diff
@@ -0,0 +1,157 @@
+ *
+ * The purpose of this test is to verify that even if there are 2 versions of
+ * the same PdxSerializable class, the right PdxType will be used during
+ * (de)serialization
+ */
+TEST(PdxSerializableTest, testRightPdxTypeForDiffPdxVersions) {
+  Cluster cluster{LocatorCount{1}, ServerCount{1}};
+  cluster.start([&]() {
+    cluster.getGfsh()
+        .deploy()
+        .jar(getFrameworkString(FrameworkVariable::JavaObjectJarPath))
+        .execute();
+  });
+  cluster.getGfsh()
+      .create()
+      .region()
+      .withName("region")
+      .withType("REPLICATE")
+      .execute();
+
+  std::shared_ptr<DeliveryAddress> entryV1;
+  std::shared_ptr<DeliveryAddress> entryV2;
+
+  {
+    auto cache = cluster.createCache();
+    auto region = setupRegion(cache);
+    cache.getTypeRegistry().registerPdxType(
+        DeliveryAddress::createDeserializable);
+
+    DeliveryAddress::setSerializationVersion(DeliveryAddress::VERSION_1);
+    entryV1 = std::make_shared<DeliveryAddress>(
+        "Some address line", "Some city", "Some country", std::string{},
+        std::shared_ptr<CacheableVector>{});
+
+    region->put("entry.v1", entryV1);
+  }
+
+  {
+    auto cache = cluster.createCache();
+    auto region = setupRegion(cache);
+    cache.getTypeRegistry().registerPdxType(
+        DeliveryAddress::createDeserializable);
+
+    DeliveryAddress::setSerializationVersion(DeliveryAddress::VERSION_1);
+    auto entry = region->get("entry.v1");
+    EXPECT_TRUE(entry);
+
+    auto address = std::dynamic_pointer_cast<DeliveryAddress>(entry);
+    EXPECT_TRUE(address);
+    EXPECT_EQ(*entryV1, *address);
+
+    DeliveryAddress::setSerializationVersion(DeliveryAddress::VERSION_2);
+    entryV2 = std::make_shared<DeliveryAddress>(
+        "Some address line", "Some city", "Some country", "Some instructions",
+        std::shared_ptr<CacheableVector>{});
+
+    auto args = CacheableVector::create();
+    args->push_back(CacheableString::create("region"));
+    args->push_back(CacheableString::create("entry.v2"));
+    args->push_back(CacheableString::create("Some address line"));
+    args->push_back(CacheableString::create("Some city"));
+    args->push_back(CacheableString::create("Some country"));
+    args->push_back(CacheableString::create("Some instructions"));
+
+    auto collector = FunctionService::onServer(region->getRegionService())
+                      .withArgs(args)
+                      .execute("PutDeliveryAddress");
+    EXPECT_TRUE(collector);
+
+    auto resultSet = collector->getResult();
+    EXPECT_TRUE(resultSet);
+
+    auto result = std::dynamic_pointer_cast<CacheableString>((*resultSet)[0]);
+    std::cerr << result->toString() << std::endl;
+  }
+
+  {
+    auto cache = cluster.createCache();
+    auto region = setupRegion(cache);
+    cache.getTypeRegistry().registerPdxType(
+        DeliveryAddress::createDeserializable);
+
+    auto entry = region->get("entry.v2");
+    EXPECT_TRUE(entry);
```

Comment:
This, and all the other `std::shared_ptr` nullability checks, should be `EXPECT_THAT(entry, IsNotNull())`.